File: docs/examples/cpp/basic.cpp
/*
 * C++ example using the RAII wrapper (tiktoken.hpp).
 */
#include "tiktoken.hpp"

#include <iostream>
#include <string>

int main() {
    try {
        tiktoken::Encoding enc("gpt2");

        const std::string text = "hello world";
        auto tokens = enc.encode_ordinary(text);

        std::cout << "tokens (" << tokens.size() << "):";
        for (uint32_t t : tokens) {
            std::cout << " " << t;
        }
        std::cout << "\n";

        std::string decoded = enc.decode(tokens);
        std::cout << "decoded: " << decoded << "\n";

        /* cl100k to compare the same text */
        tiktoken::Encoding cl100k("cl100k_base");
        auto cl_tokens = cl100k.encode_ordinary(text);
        std::cout << "cl100k tokens (" << cl_tokens.size() << "):";
        for (uint32_t t : cl_tokens) {
            std::cout << " " << t;
        }
        std::cout << "\n";

        auto names = tiktoken::list_encoding_names();
        std::cout << "available encodings: " << names.size() << "\n";

    } catch (const tiktoken::Error& e) {
        std::cerr << "error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}

File: docs/examples/rust/basic.rs
//! Rust example - direct crate usage (no C FFI).
//!
//! Run from the repository root:
//!   cargo run --example basic --manifest-path docs/examples/rust/Cargo.toml
//!
//! Or add `tiktoken = { path = "../.." }` in your Cargo.toml.

use tiktoken::get_encoding;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let enc = get_encoding("gpt2")?;

    let text = "hello world";
    let tokens = enc.encode_ordinary(text);
    println!("tokens ({}):" , tokens.len());
    for t in &tokens {
        print!(" {t}");
    }
    println!();

    let decoded = enc.decode(&tokens)?;
    println!("decoded: {decoded}");

    let allowed: std::collections::HashSet<&str> = ["<|endoftext|>"].into();
    let (eot_tokens, _) = enc.encode("hello <|endoftext|>", &allowed)?;
    println!("with eot: {eot_tokens:?}");

    println!("encodings: {:?}", tiktoken::list_encoding_names());

    Ok(())
}
